lib/foreman/renderer.rb

```diff
@@ -1,389 +1,42 @@
-        render :partial => "unattended/snippets/#{file}"
-      else
-        snippet(file.gsub(/^_/, ""), options)
+      def configure
+        yield config
       end
-    end
 
-    def snippet(name, options = {})
-      if (template = Template.where(:name => name, :snippet => true).first)
-        begin
-          return unattended_render(template, nil, options[:variables] || {})
-        rescue => exc
-          if exc.is_a?(::Foreman::Exception)
-            raise exc
-          else
-            e = ::Foreman::Exception.new(N_("The snippet '%{name}' threw an error: %{exc}"), { :name => name, :exc => exc })
-            e.set_backtrace exc.backtrace
-            raise e
-          end
-        end
-      else
-        if options[:silent]
-          nil
-        else
-          raise "The specified snippet '#{name}' does not exist, or is not a snippet."
+      def render_template_to_tempfile(template:, prefix:, host: nil, params: {}, variables: {}, options: {})
+        file = ""
+        source = get_source(template: template, host: host)
+        scope = get_scope(host: host, params: params, variables: variables)
+        Tempfile.open(prefix, Rails.root.join('tmp')) do |f|
+          f.print render(source, scope)
+          f.flush
+          f.chmod options[:mode] if options[:mode]
+          file = f
         end
+        file
       end
-    end
-
-    def snippet_if_exists(name, options = {})
-      snippet name, options.merge(:silent => true)
-    end
-
-    def save_to_file(filename, content)
-      "cat << EOF > #{filename}\n#{content}EOF"
-    end
-
-    def indent(count)
-      return unless block_given? && (text = yield.to_s)
-      prefix = " " * count
-      prefix + text.gsub(/\n/, "\n#{prefix}")
-    end
 
-    def dns_lookup(name_or_ip)
-      resolver = Resolv::DNS.new
-      Timeout.timeout(Setting[:dns_conflict_timeout]) do
-        begin
-          resolver.getname(name_or_ip)
-        rescue Resolv::ResolvError
-          resolver.getaddress(name_or_ip)
-        end
-      end
-    rescue => e
-      log_warn "Template helper dns_lookup failed: #{e}"
-      raise e
-    end
-
-    # accepts either template object or plain string
-    def unattended_render(template, overridden_name = nil, scope_variables = {})
-      @template_name = template.respond_to?(:name) ? template.name : (overridden_name || 'Unnamed')
-      template_logger.info "Rendering template '#{@template_name}'"
-      raise ::Foreman::Exception.new(N_("Template '%s' is either missing or has an invalid organization or location"), @template_name) if template.nil?
-      content = template.respond_to?(:template) ? template.template : template
-      allowed_variables = allowed_variables_mapping(ALLOWED_VARIABLES)
-      result = render_safe content, ALLOWED_HELPERS, allowed_variables, scope_variables
-      digest = Digest::SHA256.hexdigest(result)
-      Foreman::Logging.blob("Unattended render of '#{@template_name}' = '#{digest}'", result,
```

Comment:
This useful logging was removed, I think it should be restored in `BaseRenderer.render` or both renderers' render method

Reply:
Yeah we recently added this, all rendered templates go into blob logger.

Reply:
@ares @lzap What about `template_context`? Can we just set `"Foreman::Renderer::SafeModeRenderer"` or `"Foreman::Renderer::UnsafeModeRenderer"` here? If that's not enough, we'll probably have to pass the context as a parameter when calling the renderer.

Reply:
Not sure I understand the question. I know to get the template name, you need to have the source instance already. But imho you could define log_rendering method in BaseRenderer and then call it in both Safe and Unsafe remderer

Reply:
@kamils-iRonin the purpose of this field is to log if this was rendered in Host instance or UnattendedRenderer instance. But since context is now fully dynamic, let's just drop the field. It would be not obvious if this was parameter to plugin authors and it would not be set anyway = useless.